File: Implementation/c++/16_문자열압축.cpp
// 이것이 취업을 위한 코딩테스트다 325p
// https://programmers.co.kr/learn/courses/30/lessons/60057

#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(string s) {
	int answer = 999999;
	int k = 1;
	for (int k = 1; k <= s.size() / 2 + 1; k++)
	{
		string prev = s.substr(0, k);
		string result;
		int count = 1;
		for (int i = k; i < s.size(); i += k)
		{
			if (prev.compare(s.substr(i, k)) == 0)
				count += 1;
			else
			{
				if (count == 1)
					result += prev;
				else
					result += to_string(count) + prev;
				prev = s.substr(i, k);
				count = 1;
			}
		}
		if (count == 1)
			result += prev;
		else
			result += to_string(count) + prev;
		if (answer > result.size())
			answer = result.size();
	}
	return answer;
}

int main()
{
	cout << solution("a") << '\n';
}

File: Implementation/python/16_문자열압축.py
# 이것이 취업을 위한 코딩테스트다 325p
# https://programmers.co.kr/learn/courses/30/lessons/60057
# 1

def solution(s):
    answer = 1e9
    if len(s) == 1:
        return 1
    for cnt in range(1, int(len(s) / 2 + 1)):
        result_str = ""
        str_cnt = 1
        tmp = s[0 : 0 + cnt]
        for i in range(cnt, len(s), cnt):
            if tmp == s[i : i + cnt]:
                str_cnt += 1
            else:
                if str_cnt == 1:
                    result_str += tmp
                else:
                    result_str += str(str_cnt) + tmp
                tmp = s[i : i + cnt]
                str_cnt = 1
        if str_cnt == 1:
            result_str += tmp
        else:
            result_str += str(str_cnt) + tmp
        answer = min(answer, len(result_str))
    return answer

print(solution(input()))